sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CertificateValidationCheck.cs

```diff
@@ -0,0 +1,299 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class CertificateValidationCheck : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4830";
+        private const string MessageFormat = "Enable server certificate validation on this SSL/TLS connection";
+
+        private static readonly DiagnosticDescriptor rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            //Handling of += delegate syntax
+            context.RegisterSyntaxNodeActionInNonGenerated(c => CheckAddHandlerSyntax(c), SyntaxKind.AddAssignmentExpression);
+
+            //Handling of constructor parameter syntax (SslStream)
+            context.RegisterSyntaxNodeActionInNonGenerated(c => CheckConstructorParameterSyntax(c), SyntaxKind.ObjectCreationExpression);
+        }
+
+        private void CheckAddHandlerSyntax(SyntaxNodeAnalysisContext c)
+        {
+            var leftIdentifier = ((AssignmentExpressionSyntax)c.Node).Left.DescendantNodesAndSelf().OfType<IdentifierNameSyntax>().LastOrDefault();
+            var right = ((AssignmentExpressionSyntax)c.Node).Right;
+            if (leftIdentifier != null && right != null
+                && c.SemanticModel.GetSymbolInfo(leftIdentifier).Symbol is IPropertySymbol left
+                && IsValidationDelegateType(left.Type))
+            {
+                TryReportLocations(new InspectionContext(c), leftIdentifier.GetLocation(), right);
+            }
+        }
+
+        private void CheckConstructorParameterSyntax(SyntaxNodeAnalysisContext c)
+        {
+            if (c.SemanticModel.GetSymbolInfo(c.Node).Symbol is IMethodSymbol ctor)
+            {
+                CSharpMethodParameterLookup methodParamLookup = null;       //Cache, there might be more of them
+                foreach (var param in ctor.Parameters.Where(x => IsValidationDelegateType(x.Type)))
+                {
+                    methodParamLookup = methodParamLookup ?? new CSharpMethodParameterLookup((c.Node as ObjectCreationExpressionSyntax).ArgumentList, ctor);
+                    if (methodParamLookup.TryGetSymbolParameter(param, out var argument))
+                    { //For Lambda expression extract location of the parentheses only to separate them from secondary location of "true"
```

Comment:
nitpick: we normally put comments on the next line (not on the same line with `{`)